docs/datatypes/stream/zstream.md

```diff
@@ -0,0 +1,1865 @@
+Producing element 6
+Processing batch of events: Chunk(6)
+Producing element 7
+Producing element 8
+Producing element 9
+Processing batch of events: Chunk(7)
+Producing element 10
+Producing element 11
+Processing batch of events: Chunk(8,9)
+Producing element 12
+Processing batch of events: Chunk(10,11)
+Processing batch of events: Chunk(12)
+```
+
+The `ZStream#aggregateAsyncWithin` is another aggregator which takes a scheduler. This scheduler will consume all events produced by the given transducer. So the `aggregateAsyncWithin` will emit when the transducer emits or when the scheduler expires:
+
+```scala
+abstract class ZStream[-R, +E, +O] {
+  def aggregateAsyncWithin[R1 <: R, E1 >: E, P](
+    transducer: ZTransducer[R1, E1, O, P],
+    schedule: Schedule[R1, Chunk[P], Any]
+  ): ZStream[R1 with Clock, E1, P] = ???
+}
+```
+
+When we are doing I/O, batching is very important. With ZIO streams, we can create user-defined batches. It is pretty easy to do that with the `ZStream#aggregateAsyncWithin` operator. Let's see the below snippet code:
+
+```scala mdoc:invisible
+case class Record()
+val dataStream: ZStream[Any, Nothing, Record] = ZStream.repeat(Record())
+```
+
+```scala mdoc:silent:nest
+dataStream.aggregateAsyncWithin(
+   ZTransducer.collectAllN(2000),
+   Schedule.fixed(30.seconds)
+ )
+```
+
+So it will collect elements into a chunk up to 2000 elements and if we have got less than 2000 elements and 30 seconds have passed, it will pass currently collected elements down the stream whether it has collected zero, one, or 2000 elements. So this is a sort of timeout for aggregation operation. This approach aggressively favors **throughput** over **latency**. It will introduce a fixed amount of latency into a stream. We will always wait for up to 30 seconds if we haven't reached this sort of boundary value.
+
+Instead, thanks to `Schedule` we can create a much smarter **adaptive batching algorithm** that can balance between **throughput** and **latency*. So what we are doing here is that we are creating a schedule that operates on chunks of records. What the `Schedule` does is that it starts off with 30-second timeouts for as long as its input has a size that is lower than 1000, now once we see an input that has a size look higher than 1000, we will switch to a second schedule with some jittery, and we will remain with this schedule for as long as the batch size is over 1000:
+
+```scala mdoc:silent:nest
+val schedule: Schedule[Clock with Random, Chunk[Chunk[Record]], Long] =
+  // Start off with 30-second timeouts as long as the batch size is < 1000
+  Schedule.fixed(30.seconds).whileInput[Chunk[Chunk[Record]]](_.flatten.length < 100) andThen
+    // and then, switch to a shorter jittered schedule for as long as batches remain over 1000
+    Schedule.fixed(5.seconds).jittered.whileInput[Chunk[Chunk[Record]]](_.flatten.length >= 1000)
+    
+dataStream
+  .aggregateAsyncWithin(ZTransducer.collectAllN(2000), schedule)
+```
```

Comment:
Hello @iravid 
Yesterday all the time, I was working on the adaptive batching pattern which you have presented in one of your talk about ZIO streams.
I tried to rewrite and test that and and then include this nice use-case in ZIO stream documentation. But I'm not sure that this example works as what you presented.
When I debug that program, I found that in the `_.flatten.length < 100` predicate, the value of `_.flatten.length` does not always represent the batch size. Most of the time, its value is zero and sometime it is `2000`. I didn't find a situation when the size of the batch is is something between 0 and 2000. (I also jittered the upstream to make sure I have different throughout)
Could you please look at that and check if I'm right or not?

Reply:
Hey! Will try to take a look later today